zerver/webhooks/slack/view.py

```diff
@@ -1,27 +1,36 @@
 from __future__ import absolute_import
 from django.utils.translation import ugettext as _
 from django.http import HttpRequest, HttpResponse
-from zerver.lib.actions import check_send_message
+from zerver.lib.actions import check_send_message, create_stream_if_needed
 from zerver.lib.response import json_success, json_error
 from zerver.decorator import REQ, has_request_variables, api_key_only_webhook_view
 from zerver.models import UserProfile, Client
 
 
 @api_key_only_webhook_view('Slack')
 @has_request_variables
-def api_slack_webhook(request, user_profile, client, user_name=REQ(), text=REQ(), channel_name=REQ(), stream=REQ(default='slack')):
-    # type: (HttpRequest, UserProfile, Client, str, str, str, str) -> HttpResponse
+def api_slack_webhook(request, user_profile, client, user_name=REQ(), text=REQ(), channel_name=REQ(), stream=REQ(default='slack'), channels_map_to_topics=REQ(default='1')):
+    # type: (HttpRequest, UserProfile, Client, str, str, str, str, str) -> HttpResponse
 
+    channels_map_to_topics = int(channels_map_to_topics)
     sender = user_name
     message = text
     channel = channel_name
 
-    if sender is None:
-        return json_error(_("Empty sender"))
-    if message is None:
-        return json_error(_("Empty message"))
-    if channel is None:
-        return json_error(_("Empty message"))
-    subject = "%s has send you the following message on %s" % (sender, channel)
-    check_send_message(user_profile, client, "stream", [stream], subject, message)
+    info_dict = {'sender': sender, 'message': message, 'channel': channel}
+
+    for info in info_dict.keys():
+        if info_dict[info] is None:
+            error_message = "Empty "+info
+            return json_error(_(error_message))
+
+    if channels_map_to_topics:
+        subject = "channel: %s" % channel
+        zulip_message = "**%s** has sent the following message:\n `%s` " % (sender, message)
+        check_send_message(user_profile, client, "stream", [stream], subject, zulip_message)
+    else:
+        create_stream_if_needed(user_profile.realm, channel)
+        subject = "You have received a message on Slack"
```

Comment:
We can shorten this to "message from Slack", since we generally like short topics.  And we should internationalize it with `_()`.